pkg/ivd/ivd_protected_entity_type_manager.go

```diff
@@ -18,142 +18,77 @@ package ivd
 	"io"
-	"net/url"
+	"sync"
 	"time"
 )
 
-type IVDProtectedEntityTypeManager struct {
-	client    *govmomi.Client
-	vsom      *vslm.GlobalObjectManager
-	cnsClient *cns.Client
-	s3Config  astrolabe.S3Config
-	vcParams  map[string]interface{} // Save the VC configuration params
-	logger    logrus.FieldLogger
-}
-
-func NewIVDProtectedEntityTypeManagerFromConfig(params map[string]interface{}, s3Config astrolabe.S3Config,
-	logger logrus.FieldLogger) (*IVDProtectedEntityTypeManager, error) {
-	logger.Infof("Creating NewIVDProtectedEntityTypeManagerFromConfig.")
-	var vcURL url.URL
-	vcHostStr, err := GetVirtualCenterFromParamsMap(params)
-	if err != nil {
-		return nil, err
-	}
-	vcHostPortStr, err := GetPortFromParamsMap(params)
-	if err != nil {
-		return nil, err
-	}
-
-	vcURL.Scheme = "https"
-	vcURL.Host = fmt.Sprintf("%s:%s", vcHostStr, vcHostPortStr)
-	insecure, err := GetInsecureFlagFromParamsMap(params)
-	if err != nil {
-		return nil, err
-	}
-	vcUser, err := GetUserFromParamsMap(params)
-	if err != nil {
-		return nil, err
-	}
-
-	vcPassword, err := GetPasswordFromParamsMap(params)
-	if err != nil {
-		return nil, err
-	}
-	vcURL.User = url.UserPassword(vcUser, vcPassword)
-	vcURL.Path = "/sdk"
-	retVal, err := newIVDProtectedEntityTypeManagerFromURL(&vcURL, s3Config, insecure, logger)
-	if err != nil {
-		logger.Errorf("Failed to create IVDProtectedEntityTypeManager with error, %v", err)
-		return retVal, err
-	}
-	retVal.vcParams = params
-	return retVal, err
-}
-
-func newKeepAliveClient(ctx context.Context, u *url.URL, insecure bool) (*govmomi.Client, error) {
-	soapClient := soap.NewClient(u, insecure)
-	vimClient, err := vim25.NewClient(ctx, soapClient)
-	if err != nil {
-		return nil, err
-	}
-
-	vimClient.RoundTripper = session.KeepAlive(vimClient.RoundTripper, 10*time.Minute)
-
-	c := &govmomi.Client{
-		Client:         vimClient,
-		SessionManager: session.NewManager(vimClient),
-	}
+const vsphereMajor = 6
+const vSphereMinor = 7
+const disklibLib64 = "/usr/lib/vmware-vix-disklib/lib64"
 
-	// Only login if the URL contains user information.
-	if u.User != nil {
-		err = c.Login(ctx, u.User)
-		if err != nil {
-			return nil, err
-		}
-	}
+var configLoadLock sync.Mutex
 
-	return c, nil
+type IVDProtectedEntityTypeManager struct {
+	vcenterConfig  *vsphere.VirtualCenterConfig
+	vcenterManager vsphere.VirtualCenterManager
+	vslmManager    vsphere.VslmManager
+	cnsManager     vsphere.CnsManager
+	s3Config       astrolabe.S3Config
+	logger         logrus.FieldLogger
 }
 
-//TODO - merge this into NewIVDProtectedEntityTypeManagerFromConfig and switch usage
-func newIVDProtectedEntityTypeManagerFromURL(url *url.URL, s3Config astrolabe.S3Config, insecure bool, logger logrus.FieldLogger) (*IVDProtectedEntityTypeManager, error) {
+func NewIVDProtectedEntityTypeManager(params map[string]interface{},
+	s3Config astrolabe.S3Config,
+	logger logrus.FieldLogger) (*IVDProtectedEntityTypeManager, error) {
+	configLoadLock.Lock()
```

Comment:
If possible, this should go through ReloadConfig.  That way you don't have to worry about keeping your init and reload code in sync.

Reply:
Actually they do different things, reload also verifies if the current vc config is different compared to the previous one.

Reply:
Right - so if your current config is nil you just go ahead and reload, right?

Reply:
Done.